### PR TITLE
Add lazy flattening function to Relation

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Adds a `flatten` function to produce and iterator from a `Relation` (i.e. a `Map[A, Set[B]]`) that is each element in the set value paired with the corresponding key. Inverse of grouping.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala [hunk 1]
```diff
@@ -35,6 +35,12 @@ object Relation {
 
       mutMap.toMap
     }
+
+    def flatten[A, B](relation: Relation[A, B]): Iterator[(A, B)] =
+      for {
+        key <- relation.keysIterator
```
Comment (S11001001) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]:
```suggestion
        kvs <- relation.iterator
        value <- relation._2
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala [hunk 2]
```diff
@@ -35,6 +35,12 @@ object Relation {
...
+        value <- relation(key)
```
Comment (stefanobaghino-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]:
```suggestion
        (key, values) <- relation.iterator
        value <- values
```
@S11001001 What about this?

Comment (S11001001) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]: Don't do that, it adds a `withFilter` call.

Comment (S11001001) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]: Yes, even though it looks innocent. You can't even write `kvs: (A, Set[B]) <-` without it adding a `withFilter`; the only safe things are variable bindings and `_`.

Comment (stefanobaghino-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]: I thought `withFilter` is only applied if a guard is in the `for`-comprehension? Can you explain, please?

Comment (S11001001) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]: A pattern counts as a guard (passing iff the pattern matches), and anything counts as a pattern for this purpose except the two exceptions I mentioned above, inducing a `withFilter` even if they appear irrefutable.

Comment (stefanobaghino-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]: Sure, the only issue with the two lines above is that they don't compile. Can I call `._2` on `kvs` or are there caveats on that?

Comment (stefanobaghino-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]: I'll assume there are no further caveats.

Comment (stefanobaghino-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]: Addressed by https://github.com/digital-asset/daml/pull/4922/commits/2c2c06646ac38fdd9939fd925f74256f54b46f95

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala]: Addressed by https://github.com/digital-asset/daml/pull/4922/commits/2c2c06646ac38fdd9939fd925f74256f54b46f95

**Pull request merged**